pkg/api/validate/dynamic/diskencryptionset.go

```diff
@@ -0,0 +1,81 @@
+package dynamic
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+)
+
+func (dv *dynamic) ValidateDiskEncryptionSets(ctx context.Context, oc *api.OpenShiftCluster) error {
```

Comment:
I think we need to validate location too. If I recall right keyvault + DES has to be in the same region too where cluster is.

Reply:
You are right:

> All resources related to your customer-managed keys (Azure Key Vaults, disk encryption sets, VMs, disks, and snapshots) must be in the same subscription and region.

However, I think it should be enough to validation only location of DES here: we do not work with KV directly and do not even need permissions for it. Sounds like validation of location of KV is a responsibility of compute RP: you first create a KV, then you create a DES via compute RP which should have own validation and only after this you create an ARO cluster providing DES resoruce ID.

Looking into it.

Reply:
@mjudeikis I added location validation.

> I think it should be enough to validation only location of DES here: we do not work with KV directly and do not even need permissions for it. Sounds like validation of location of KV is a responsibility of compute RP...

I looked into it and it seems like compute RP doesn't validate location of the KV when creating DES: I was able to create a DES in westeurope with a KV from eastus. Not sure whether it is just outdated documentation or lack of validation on the compute RP.

I would prefer not to validate KV location, becuase it will likely require extra permissions for RP and cluster SP. But we can look into it. What do you think?

Reply:
@mjudeikis, ok so it looks like compute RP doesn't validate when you create a DES, but does validate when you create a VM. I got a `DeploymentFailed` with `KeyVaultAndDiskInDifferentRegions` for each master and bootstrap VM.

Not great. I wish they were validating it during DES creation...

Reply:
Now this is something I seen :) we need to catch this

Reply:
I looked into catching it and there are two options:

1. Catch in the dynamic validation.
2. Catch `DeploymentFailed` and check if one of the deploymnet errors has something to do with `KeyVaultAndDiskInDifferentRegions`.

And both end up being a mess.

If we want to catch on validation we need:

1. Customer to give us extra permissions which is not great as we only need it during cluster creation for validation of something which is actually not ARO responsibility.
2. We will have to do the following (additional) checks:
    * Does disk encrhption set have an a source vault set (not sure in which circumstances it can be `nil`, but the field is optional/`nil`able)
    * Do we have permissions to read KV object?
    * Does the key vault actually exist (you can create a KV & disk encryption set, then delete key vault and try to create a cluster)
    * Is KV in the same region as cluster?
    * Probably some other edge cases.

If we want to catch this on `DeploymentFailed` - then it is also tricky. We already catching `DeploymentFailed` in [`pkg/cluster/arm.go`](https://github.com/Azure/ARO-RP/blob/68ab4822c57114e29af7802a0dab02ab9895948d/pkg/cluster/arm.go) and propogating it to a customer. If we want to have a nice message then, we need to handle at least these failures:

1. Disk encryption set exists and in correct region, but KV is in different region. We can catch it by `KeyVaultAndDiskInDifferentRegions`.
2. Disk encryption set exists, but key vault doesn't exist. We can't catch it easily: errors are just `BadRequest` and text `message`.

Even if we do catch these cases, what do we do next? Do we introduce a special case of `api.CloudErrorCodeDeploymentFailed` where we say "your key vault is in different region/doesn't exist"? What about other errors from `DeploymentFailed`? It is possible that DES/KV errors are not the only cause of deploymnet being failed. Do we just ignore other errors?

After looking and trying different options, my opition on this is - it is not worth handling as we already handling `DeploymentFailed` in [`pkg/cluster/arm.go`](https://github.com/Azure/ARO-RP/blob/68ab4822c57114e29af7802a0dab02ab9895948d/pkg/cluster/arm.go) and propogating all the erors from ARM to the customer:

https://github.com/Azure/ARO-RP/blob/68ab4822c57114e29af7802a0dab02ab9895948d/pkg/cluster/arm.go#L47-L62